src/Models/UserIdentityModel.php

```diff
@@ -25,6 +26,146 @@ class UserIdentityModel extends Model
     ];
     protected $useTimestamps = true;
 
+    public function getAccessTokenByToken(string $token): ?AccessToken
+    {
+        return $this
+            ->where('type', 'access_token')
+            ->where('secret', hash('sha256', $token))
+            ->asObject(AccessToken::class)
+            ->first();
+    }
+
+    /**
+     * @param int|string $userId
+     */
+    public function getAccessToken($userId, string $token): ?AccessToken
+    {
+        return $this->where('user_id', $userId)
+            ->where('type', 'access_token')
+            ->where('secret', hash('sha256', $token))
+            ->asObject(AccessToken::class)
+            ->first();
+    }
+
+    /**
+     * Given the ID, returns the given access token.
+     *
+     * @param int|string $id
+     * @param int|string $userId
+     */
+    public function getAccessTokenById($id, $userId): ?AccessToken
+    {
+        return $this->where('user_id', $userId)
+            ->where('type', 'access_token')
+            ->where('id', $id)
+            ->asObject(AccessToken::class)
+            ->first();
+    }
+
+    /**
+     * @param int|string $userId
+     *
+     * @return AccessToken[]
+     */
+    public function getAllAccessTokens($userId): array
+    {
+        return $this
+            ->where('user_id', $userId)
+            ->where('type', 'access_token')
+            ->asObject(AccessToken::class)
+            ->find();
+    }
+
+    /**
+     * @param int|string $userId
+     */
+    public function getIdentityByType($userId, string $type): ?UserIdentity
+    {
+        return $this->where('user_id', $userId)
+            ->where('type', $type)
+            ->first();
```

Comment:
Why not consolidate these?
```suggestion
        return reset($this->getIdentitiesByTypes($userId, [$type]));
```

Reply:
Can't do that, because here is `first()`.

Reply:
They should be functionally equivalent since they will end up using the same indexes. Is it a performance concern, loading all identities then returning just the first?

Reply:
Right now it does not make sense to shorten the code of a small method. 
It is also not yet certain that the methods we have created here are really necessary.

We are still in the process of modifying and finalizing the api.